File: engine/src/render/charts.rs
//! Chart.js HTML snippet generators. Each function returns self-contained HTML
//! with an inline Chart.js CDN script tag and canvas element.

use uuid::Uuid;

const CHARTJS_CDN: &str = "https://cdn.jsdelivr.net/npm/chart.js@4.4.7/dist/chart.umd.min.js";

/// Unique DOM id for a chart canvas.
fn chart_id() -> String {
    format!("chart-{}", &Uuid::new_v4().to_string()[..8])
}

/// Escape a string for safe embedding in a JSON value inside an HTML `<script>` tag.
fn json_escape(s: &str) -> String {
    s.replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
        .replace('<', "\\u003c")
        .replace('>', "\\u003e")
}

/// Format a slice of f64 as a JSON array literal.
fn values_json(values: &[f64]) -> String {
    let inner: Vec<String> = values.iter().map(|v| format!("{v}")).collect();
    format!("[{}]", inner.join(","))
}

/// Format a string slice as a JSON array of strings.
fn labels_json(labels: &[String]) -> String {
    let inner: Vec<String> = labels
        .iter()
        .map(|l| format!("\"{}\"", json_escape(l)))
        .collect();
    format!("[{}]", inner.join(","))
}

/// Generate a self-contained bar chart HTML snippet.
///
/// `color` is a CSS color string (e.g. `"#a78bfa"`).
pub fn bar_chart(title: &str, labels: &[String], values: &[f64], color: &str) -> String {
    let cid = chart_id();
    let labels_js = labels_json(labels);
    let values_js = values_json(values);
    let title_esc = json_escape(title);
    let color_esc = json_escape(color);
    let show_title = !title.is_empty();

    let mut s = String::new();
    s.push_str(&format!("<script src=\"{}\"></script>\n", CHARTJS_CDN));
    s.push_str("<div style=\"max-width:100%;height:300px\">\n");
    s.push_str(&format!("  <canvas id=\"{}\"></canvas>\n", cid));
    s.push_str("</div>\n");
    s.push_str("<script>\n");
    s.push_str(&format!("new Chart(document.getElementById(\"{}\"), {{\n", cid));
    s.push_str("  type: \"bar\",\n");
    s.push_str(&format!("  data: {{\n    labels: {},\n", labels_js));
    s.push_str(&format!(
        "    datasets: [{{\n      label: \"{}\",\n      data: {},\n",
        title_esc, values_js
    ));
    s.push_str(&format!(
        "      backgroundColor: \"{}40\",\n      borderColor: \"{}\",\n      borderWidth: 2\n    }}]\n",
        color_esc, color_esc
    ));
    s.push_str("  },\n  options: {\n    responsive: true,\n");
    s.push_str(&format!(
        "    plugins: {{ title: {{ display: {}, text: \"{}\", color: \"#e2e8f0\" }} }},\n",
        show_title, title_esc
    ));
    s.push_str("    scales: {\n");
    s.push_str(
        "      x: { ticks: { color: \"#94a3b8\" }, grid: { color: \"#1e293b\" } },\n",
    );
    s.push_str(
        "      y: { ticks: { color: \"#94a3b8\" }, grid: { color: \"#1e293b\" } }\n",
    );
    s.push_str("    }\n  }\n});\n</script>");
    s
}

/// Generate a self-contained line chart HTML snippet.
///
/// Each dataset tuple is `(label, values, color)`.
pub fn line_chart(
    title: &str,
    labels: &[String],
    datasets: &[(String, Vec<f64>, String)],
) -> String {
    let cid = chart_id();
    let labels_js = labels_json(labels);
    let title_esc = json_escape(title);
    let show_title = !title.is_empty();

    let datasets_js: Vec<String> = datasets
        .iter()
        .map(|(label, data, color)| {
            format!(
                "{{label:\"{}\",data:{},borderColor:\"{}\",backgroundColor:\"{}20\",fill:false,tension:0.3,borderWidth:2,pointRadius:3}}",
                json_escape(label),
                values_json(data),
                json_escape(color),
                json_escape(color),
            )
        })
        .collect();

    let mut s = String::new();
    s.push_str(&format!("<script src=\"{}\"></script>\n", CHARTJS_CDN));
    s.push_str("<div style=\"max-width:100%;height:300px\">\n");
    s.push_str(&format!("  <canvas id=\"{}\"></canvas>\n", cid));
    s.push_str("</div>\n");
    s.push_str("<script>\n");
    s.push_str(&format!("new Chart(document.getElementById(\"{}\"), {{\n", cid));
    s.push_str("  type: \"line\",\n");
    s.push_str(&format!("  data: {{\n    labels: {},\n", labels_js));
    s.push_str(&format!("    datasets: [{}]\n", datasets_js.join(",")));
    s.push_str("  },\n  options: {\n    responsive: true,\n");
    s.push_str(&format!(
        "    plugins: {{ title: {{ display: {}, text: \"{}\", color: \"#e2e8f0\" }} }},\n",
        show_title, title_esc
    ));
    s.push_str("    scales: {\n");
    s.push_str(
        "      x: { ticks: { color: \"#94a3b8\" }, grid: { color: \"#1e293b\" } },\n",
    );
    s.push_str(
        "      y: { ticks: { color: \"#94a3b8\" }, grid: { color: \"#1e293b\" } }\n",
    );
    s.push_str("    }\n  }\n});\n</script>");
    s
}

/// Generate a self-contained pie/doughnut chart HTML snippet.
///
/// `colors` should contain one CSS color per value.
pub fn pie_chart(title: &str, labels: &[String], values: &[f64], colors: &[String]) -> String {
    let cid = chart_id();
    let labels_js = labels_json(labels);
    let values_js = values_json(values);
    let title_esc = json_escape(title);
    let show_title = !title.is_empty();

    let bg: Vec<String> = colors
        .iter()
        .map(|c| format!("\"{}80\"", json_escape(c)))
        .collect();
    let border: Vec<String> = colors
        .iter()
        .map(|c| format!("\"{}\"", json_escape(c)))
        .collect();

    let mut s = String::new();
    s.push_str(&format!("<script src=\"{}\"></script>\n", CHARTJS_CDN));
    s.push_str("<div style=\"max-width:100%;height:300px\">\n");
    s.push_str(&format!("  <canvas id=\"{}\"></canvas>\n", cid));
    s.push_str("</div>\n");
    s.push_str("<script>\n");
    s.push_str(&format!("new Chart(document.getElementById(\"{}\"), {{\n", cid));
    s.push_str("  type: \"doughnut\",\n");
    s.push_str(&format!("  data: {{\n    labels: {},\n", labels_js));
    s.push_str(&format!(
        "    datasets: [{{\n      data: {},\n      backgroundColor: [{}],\n      borderColor: [{}],\n      borderWidth: 2\n    }}]\n",
        values_js,
        bg.join(","),
        border.join(","),
    ));
    s.push_str("  },\n  options: {\n    responsive: true,\n");
    s.push_str(&format!(
        "    plugins: {{\n      title: {{ display: {}, text: \"{}\", color: \"#e2e8f0\" }},\n      legend: {{ labels: {{ color: \"#94a3b8\" }} }}\n    }}\n",
        show_title, title_esc
    ));
    s.push_str("  }\n});\n</script>");
    s
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_bar_chart_contains_canvas() {
        let html = bar_chart(
            "Revenue",
            &["Q1".into(), "Q2".into()],
            &[100.0, 200.0],
            "#a78bfa",
        );
        assert!(html.contains("<canvas"));
        assert!(html.contains("chart.js"));
        assert!(html.contains("Revenue"));
        assert!(html.contains("bar"));
    }

    #[test]
    fn test_line_chart_multiple_datasets() {
        let datasets = vec![
            ("A".into(), vec![1.0, 2.0, 3.0], "#22c55e".into()),
            ("B".into(), vec![3.0, 2.0, 1.0], "#ef4444".into()),
        ];
        let html = line_chart(
            "Trends",
            &["Jan".into(), "Feb".into(), "Mar".into()],
            &datasets,
        );
        assert!(html.contains("<canvas"));
        assert!(html.contains("line"));
        assert!(html.contains("#22c55e"));
        assert!(html.contains("#ef4444"));
    }

    #[test]
    fn test_pie_chart_colors() {
        let html = pie_chart(
            "Split",
            &["X".into(), "Y".into()],
            &[60.0, 40.0],
            &["#a78bfa".into(), "#22c55e".into()],
        );
        assert!(html.contains("doughnut"));
        assert!(html.contains("#a78bfa"));
        assert!(html.contains("#22c55e"));
    }

    #[test]
    fn test_bar_chart_empty_title() {
        let html = bar_chart("", &["A".into()], &[1.0], "#000");
        assert!(html.contains("display: false"));
    }
}

File: legacy/src/datamirai_engine/render/charts.py
"""Chart helpers — generate Chart.js chart configurations as HTML snippets.

Each helper returns a self-contained <canvas> + <script> block that renders
a chart using Chart.js (loaded inline in the HTML document by render_rich).
"""

from __future__ import annotations

import json
import uuid


def _chart_id() -> str:
    return f"chart-{uuid.uuid4().hex[:8]}"


def bar_chart(
    labels: list[str],
    datasets: list[dict],
    title: str = "",
    height: int = 300,
) -> str:
    """Generate a bar chart HTML snippet.

    datasets: list of {label: str, data: list[number], color?: str}
    """
    cid = _chart_id()
    ds = []
    colors = ["#a78bfa", "#22c55e", "#06b6d4", "#f59e0b", "#ef4444", "#ec4899"]
    for i, d in enumerate(datasets):
        color = d.get("color", colors[i % len(colors)])
        ds.append({
            "label": d.get("label", f"Serie {i+1}"),
            "data": d["data"],
            "backgroundColor": color + "40",
            "borderColor": color,
            "borderWidth": 2,
        })
    config = {
        "type": "bar",
        "data": {"labels": labels, "datasets": ds},
        "options": {
            "responsive": True,
            "plugins": {"title": {"display": bool(title), "text": title, "color": "#e2e8f0"}},
            "scales": {
                "x": {"ticks": {"color": "#94a3b8"}, "grid": {"color": "#1e293b"}},
                "y": {"ticks": {"color": "#94a3b8"}, "grid": {"color": "#1e293b"}},
            },
        },
    }
    return f'<div style="max-width:100%;height:{height}px"><canvas id="{cid}"></canvas></div>\n<script>new Chart(document.getElementById("{cid}"),{json.dumps(config)});</script>'


def line_chart(
    labels: list[str],
    datasets: list[dict],
    title: str = "",
    height: int = 300,
) -> str:
    """Generate a line chart HTML snippet."""
    cid = _chart_id()
    ds = []
    colors = ["#a78bfa", "#22c55e", "#06b6d4", "#f59e0b", "#ef4444", "#ec4899"]
    for i, d in enumerate(datasets):
        color = d.get("color", colors[i % len(colors)])
        ds.append({
            "label": d.get("label", f"Serie {i+1}"),
            "data": d["data"],
            "borderColor": color,
            "backgroundColor": color + "20",
            "fill": d.get("fill", False),
            "tension": 0.3,
            "borderWidth": 2,
            "pointRadius": 3,
        })
    config = {
        "type": "line",
        "data": {"labels": labels, "datasets": ds},
        "options": {
            "responsive": True,
            "plugins": {"title": {"display": bool(title), "text": title, "color": "#e2e8f0"}},
            "scales": {
                "x": {"ticks": {"color": "#94a3b8"}, "grid": {"color": "#1e293b"}},
                "y": {"ticks": {"color": "#94a3b8"}, "grid": {"color": "#1e293b"}},
            },
        },
    }
    return f'<div style="max-width:100%;height:{height}px"><canvas id="{cid}"></canvas></div>\n<script>new Chart(document.getElementById("{cid}"),{json.dumps(config)});</script>'


def pie_chart(
    labels: list[str],
    data: list[float],
    title: str = "",
    height: int = 300,
) -> str:
    """Generate a pie/doughnut chart HTML snippet."""
    cid = _chart_id()
    colors = ["#a78bfa", "#22c55e", "#06b6d4", "#f59e0b", "#ef4444", "#ec4899", "#8b5cf6", "#14b8a6"]
    bg = [colors[i % len(colors)] + "80" for i in range(len(data))]
    border = [colors[i % len(colors)] for i in range(len(data))]
    config = {
        "type": "doughnut",
        "data": {
            "labels": labels,
            "datasets": [{"data": data, "backgroundColor": bg, "borderColor": border, "borderWidth": 2}],
        },
        "options": {
            "responsive": True,
            "plugins": {
                "title": {"display": bool(title), "text": title, "color": "#e2e8f0"},
                "legend": {"labels": {"color": "#94a3b8"}},
            },
        },
    }
    return f'<div style="max-width:100%;height:{height}px"><canvas id="{cid}"></canvas></div>\n<script>new Chart(document.getElementById("{cid}"),{json.dumps(config)});</script>'


# CDN URL for Chart.js (loaded inline)
CHARTJS_CDN = "https://cdn.jsdelivr.net/npm/chart.js@4.4.7/dist/chart.umd.min.js"

# Minimal inline loader script
CHARTJS_LOADER = f'<script src="{CHARTJS_CDN}"></script>'
